### PR TITLE
Added new upcoming project tasks widget [IT508]

### PR DESCRIPTION
**Pull request by konadrian**
Added new upcoming project tasks widget and completed task widget [IT508]

image

### CODE REVIEW

File: public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js [hunk 1]
```diff
@@ -2921,4 +2921,35 @@ YetiForce_Widget_Js('YetiForce_Multifilter_Widget_Js', {}, {
...
+			let url = 'index.php?module=ProjectTask&view=List&viewname=All';
+			url += '&search_params=[[';
+			let owner = container.find('.widgetFilter.owner option:selected');
+			if (owner.val() !== 'all') {
+				url += '["assigned_user_id","e","' + owner.val() + '"],';
+			}
+			let status;
+			let activitiesStatus = container.data('name');
+			if (activitiesStatus === 'UpcomingProjectTasks') {
+				status = 'PLL_PLANNED##PLL_ON_HOLD##PLL_SUBMITTED_COMMENTS##PLL_IN_PROGRESSING##PLL_IN_APPROVAL';
```
Comment (TomaszQr) [inline public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js]: statuses can be different

File: public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js [hunk 2]
```diff
@@ -2921,4 +2921,35 @@ YetiForce_Widget_Js('YetiForce_Multifilter_Widget_Js', {}, {
...
+			} else {
+				status = 'PLL_COMPLETED##PLL_CANCELLED';
+			}
+			url += '["projecttaskstatus","e","' + encodeURIComponent(status) + '"]]]';
+			window.location.href = url;
```
Comment (TomaszQr) [inline public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js]: app.openUrl

File: modules/ProjectTask/models/Module.php [hunk 3]
```diff
@@ -1,13 +1,61 @@
...
+	/**
+	 * Get project tasks by status.
+	 *
+	 * @param array  $params
+	 * @param object $pagingModel
+	 * @param mixed  $user
+	 *
+	 * @return array
+	 */
+	public static function getRecordsByStatus(array $params, object $pagingModel, $user): array
```
Comment (TomaszQr) [inline modules/ProjectTask/models/Module.php]: object ?

File: app/Fields/Picklist.php [hunk 4]
```diff
@@ -350,4 +350,21 @@ public static function getCloseStates(int $tabId, bool $byName = true)
...
+	 *  Get picklist values by automation value.
+	 *
+	 * @param string $fieldName
+	 * @param int    $automation
+	 *
+	 * @return array
+	 */
+	public static function getValuesByAutomation(string $fieldName, int $automation = 0): array
+	{
+		if ((bool) \App\Db::getInstance()->getTableSchema("vtiger_$fieldName", true)->getColumn('automation')) {
```
Comment (TomaszQr) [inline app/Fields/Picklist.php]: Add cache

File: modules/ProjectTask/dashboards/CompletedProjectTasks.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
+<?php
+/**
+ * CompletedProjectTasks chart class.
+ *
```
Comment (mariuszkrzaczkowski) [inline modules/ProjectTask/dashboards/CompletedProjectTasks.php]: @package

File: modules/ProjectTask/dashboards/CompletedProjectTasks.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+class ProjectTask_CompletedProjectTasks_Dashboard extends Vtiger_IndexAjax_View
+{
+	/**
+	 * Process.
+	 *
+	 * @param \App\Request $request
+	 */
+	public function process(\App\Request $request)
+	{
+		$currentUser = App\User::getCurrentUserModel();
```
Comment (mariuszkrzaczkowski) [inline modules/ProjectTask/dashboards/CompletedProjectTasks.php]: \App\User::getCurrentUserId()

File: modules/ProjectTask/dashboards/UpcomingProjectTasks.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
+<?php
+/**
+ * UpcomingProjectTasks chart class.
+ *
```
Comment (mariuszkrzaczkowski) [inline modules/ProjectTask/dashboards/UpcomingProjectTasks.php]: @Package

File: modules/ProjectTask/dashboards/UpcomingProjectTasks.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * UpcomingProjectTasks class.
+ */
+class ProjectTask_UpcomingProjectTasks_Dashboard extends Vtiger_IndexAjax_View
+{
+	/**
+	 * Process.
+	 *
+	 * @param \App\Request $request
```
Comment (mariuszkrzaczkowski) [inline modules/ProjectTask/dashboards/UpcomingProjectTasks.php]: /**
	 * {@inheritdoc}
	 */

File: modules/ProjectTask/dashboards/CompletedProjectTasks.php [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * CompletedProjectTasks class.
+ */
+class ProjectTask_CompletedProjectTasks_Dashboard extends Vtiger_IndexAjax_View
+{
+	/**
+	 * Process.
+	 *
+	 * @param \App\Request $request
```
Comment (mariuszkrzaczkowski) [inline modules/ProjectTask/dashboards/CompletedProjectTasks.php]: /**
	 * {@inheritdoc}
	 */

File: modules/ProjectTask/dashboards/UpcomingProjectTasks.php [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+	 */
+	public function process(\App\Request $request)
+	{
+		$currentUser = App\User::getCurrentUserModel();
```
Comment (mariuszkrzaczkowski) [inline modules/ProjectTask/dashboards/UpcomingProjectTasks.php]: \App\User::getCurrentUserId()

File: modules/ProjectTask/models/Module.php [hunk 11]
```diff
@@ -1,13 +1,61 @@
...
+	 * @param Vtiger_Paging_Model $pagingModel
+	 * @param mixed               $user
+	 *
+	 * @return array
+	 */
+	public static function getRecordsByStatus(array $params, Vtiger_Paging_Model $pagingModel, $user): array
+	{
+		$query = new \App\Db\Query();
+		if (!$user) {
+			$user = (int) App\User::getCurrentUserModel()->getId();
```
Comment (mariuszkrzaczkowski) [inline modules/ProjectTask/models/Module.php]: \App\User::getCurrentUserId()

File: modules/ProjectTask/models/Module.php [hunk 12]
```diff
@@ -1,13 +1,61 @@
...
+		}
+		$query->select(['vtiger_crmentity.crmid', 'vtiger_crmentity.smownerid', 'vtiger_crmentity.setype', 'vtiger_projecttask.*'])
```
Comment (mariuszkrzaczkowski) [inline modules/ProjectTask/models/Module.php]: QueryGenerator

**Pull request merged**